pkgs/build-support/buildenv/builder.pl

```diff
@@ -14,14 +14,44 @@
 
 my @pathsToLink = split ' ', $ENV{"pathsToLink"};
 
-sub isInPathsToLink {
+# Normalize paths in pathsToLink. Normalized paths do not have duplicated
+# or trailing slashes and have a leading slash except for the root which is
+# an empty string. E.g.: "", "/foo", "/foo/bar".
+foreach my $pathToLink (@pathsToLink) {
+    my @parts = split(/\//, $pathToLink);
+    @parts = grep { $_ ne '' } @parts;
+    $pathToLink = join "", map { "/" . $_ } @parts;
+}
+
+# Determine if two paths are related, that is they are equal or one is below
+# the other. Both paths must be normalized as done above for pathsToLink
+# (note that $relName in findFiles is normalized due to its construction).
+# Examples: ("/a/b", "") -> 1, ("/a/b", "/a") -> 1, ("/a/b", "/a/b") -> 1,
+# ("/a/b", "/a/b/c") -> 1, ("/a/b", "/c") -> 0, ("/a/b", "/c/d/e") -> 0.
+sub pathsAreRelated {
+    my ($path1, $path2) = @_;
+    my $len1 = length($path1);
+    my $len2 = length($path2);
+    if ($len1 == $len2) {
+        return $path1 eq $path2;
```

Comment:
Correct me if I'm wrong, but if I see this correctly, this problem won't appear in our implementation as `$path2` is always an element from `$pathsToLink`. So if `$path1` is related to `$path2` (the path to link) and has the same size as `$path2`, your scenario shouldn't appear, right? At least that's how I understood the code and interpreted `Determine if a path is related to any of the paths in pathsToLink` :)

Reply:
Given `/a/b/xxxxx` and `/a/b/c/xxx`, the function will indeed return false. And this is correct because the paths are not related, i.e. one is not above/below the other. See that they diverge at `xxxxx` and `c`. It is easy to see if you imagine it as path/graph. This is consistent with the description, which says the paths are related when "they are equal or one is below the other". These paths are not equal, and neither is below the other.

As for why this is the right behavior. Consider that `pathsToLink` is equal to [`/a/b/c/xxx`]. Paths related to this path (the processing of which will not be skipped) are:
- The paths above this path: "", `/a`, `/a/b`, `/a/b/c` (for this case these are all such paths).
- This path itself: `/a/b/c/xxx`.
- Paths below this path, e.g. `/a/b/c/xxx/y`, `/a/b/c/xxx/y/z`... (there is an infinite number of these).

All of these related paths must not be skipped during recursive processing as that would result in skipping paths that must not be skipped since they are "covered" by `pathsToLink`. The first point is especially important because e.g. if you skip `/a` then you effectively also skip `/a/b/c/xxx`.

Paths that are **not related** to `/a/b/c/xxx` can safely be skipped during processing, such as `/a/b/xxxxx`. Just think about it, if you only want to include `/a/b/c/xxx`, then you do not want `/a/b/xxxxx` or anything under that.

@Ma27 Sorry I don't understand what you mean. As I explained, I don't see any incorrect behavior :)

Reply:
> @Ma27 Sorry I don't understand what you mean. As I explained, I don't see any incorrect behavior :)

Well, I actually agree with you, I just didn't explain this as good as you did :D 

I probably misinterpreted aszlig's comment as I guessed that with the `xxx{,xx}` being a file the function **might** be supposed to return true if the directory paths match and I tried to explain that this won't happen as `$path2` is a directory path from `pathsToLink`.

So I guess we mean the same (a.k.a. that the current implementation is appropriate), my previous comment wasn't sufficiently well-phrased :)